pkg/assembler/backends/testing/artifact.go

```diff
@@ -17,67 +17,154 @@ package testing
+	occurrences []uint32
+}
+
+func (n *artStruct) getID() uint32 { return n.id }
+
+func registerAllArtifacts(c *demoClient) {
+	c.IngestArtifact(context.Background(), &model.ArtifactInputSpec{
+		Algorithm: "sha256",
+		Digest:    "6bbb0da1891646e58eb3e6a63af3a6fc3c8eb5a0d44824cba581d2e14a0450cf",
+	})
+	c.IngestArtifact(context.Background(), &model.ArtifactInputSpec{
+		Algorithm: "sha1",
+		Digest:    "7A8F47318E4676DACB0142AFA0B83029CD7BEFD9",
+	})
+	c.IngestArtifact(context.Background(), &model.ArtifactInputSpec{
+		Algorithm: "sha512",
+		Digest:    "374AB8F711235830769AA5F0B31CE9B72C5670074B34CB302CDAFE3B606233EE92EE01E298E5701F15CC7087714CD9ABD7DDB838A6E1206B3642DE16D9FC9DD7",
+	})
 }
 
 // Ingest Artifacts
 
-func (c *demoClient) registerArtifact(algorithm, digest string) *model.Artifact {
-	// enforce lowercase for both the algorithm and digest when ingesting
-	lowerCaseDigest := strings.ToLower(digest)
-	lowerCaseAlgorithm := strings.ToLower(algorithm)
+func (c *demoClient) IngestArtifact(ctx context.Context, artifact *model.ArtifactInputSpec) (*model.Artifact, error) {
+	algorithm := strings.ToLower(artifact.Algorithm)
+	digest := strings.ToLower(artifact.Digest)
+	a, err := c.artifactByKey(algorithm, digest)
 
-	for _, a := range c.artifacts {
-		if a.Digest == lowerCaseDigest && a.Algorithm == lowerCaseAlgorithm {
-			return a
+	if err != nil {
+		a = &artStruct{
+			id:        c.getNextID(),
+			algorithm: algorithm,
+			digest:    digest,
 		}
+		c.index[a.id] = a
+		c.artifacts[strings.Join([]string{algorithm, digest}, ":")] = a
+	}
+
+	return convArtifact(a), nil
+}
+
+func (c *demoClient) artifactByID(id uint32) (*artStruct, error) {
+	o, ok := c.index[id]
+	if !ok {
+		return nil, errors.New("could not find artifact")
 	}
-	newArtifact := &model.Artifact{
-		Digest:    lowerCaseDigest,
-		Algorithm: lowerCaseAlgorithm,
+	a, ok := o.(*artStruct)
+	if !ok {
+		return nil, errors.New("not an artifact")
 	}
-	c.artifacts = append(c.artifacts, newArtifact)
+	return a, nil
+}
 
-	return newArtifact
+func (c *demoClient) artifactByKey(alg, dig string) (*artStruct, error) {
+	algorithm := strings.ToLower(alg)
+	digest := strings.ToLower(dig)
+	if a, ok := c.artifacts[strings.Join([]string{algorithm, digest}, ":")]; ok {
+		return a, nil
+	}
+	return nil, errors.New("artifact not found")
+}
+
+func (c *demoClient) artifactExact(artifactSpec *model.ArtifactSpec) (*artStruct, error) {
+	algorithm := strings.ToLower(nilToEmpty(artifactSpec.Algorithm))
+	digest := strings.ToLower(nilToEmpty(artifactSpec.Digest))
+
+	// If ID is provided, try to look up, then check if algo and digest match.
+	if artifactSpec.ID != nil {
+		id64, err := strconv.ParseUint(*artifactSpec.ID, 10, 32)
+		if err != nil {
+			return nil, fmt.Errorf("couldn't parse id %w", err)
+		}
+		id := uint32(id64)
+		a, err := c.artifactByID(id)
+		if err != nil {
+			// Not found
+			return nil, nil
+		}
+		if algorithm != "" && algorithm != a.algorithm {
```

Comment:
yeah I dont think we need to do more matching if the nodeID is provided.

Reply:
hmm, my comment didn't save :(

I think it depends on what we want to do in this corner case, but most databases will continue to match and return 0 rows/results on queries such as

```
select x
where x.id == 42 and x.stuff = foo ...
```

(assuming `.stuff` does not match).

There likely is not a performance degradation to continue matching, since we only have exactly one result.

Reply:
Updated to not match other fields if ID is provided.